File: app/src/main/java/com/adiaz/deportesmadrid/activities/DistrictActivity.java
package com.adiaz.deportesmadrid.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;

import com.adiaz.deportesmadrid.R;
import com.adiaz.deportesmadrid.adapters.GenericAdapter;
import com.adiaz.deportesmadrid.db.daos.CompetitionsDAO;
import com.adiaz.deportesmadrid.db.entities.Competition;
import com.adiaz.deportesmadrid.utils.Constants;
import com.adiaz.deportesmadrid.utils.ListItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DistrictActivity extends AppCompatActivity implements GenericAdapter.ListItemClickListener {


    //private static final String TAG = DistrictActivity.class.getSimpleName();

    @BindView(R.id.pb_loading_districts)
    ProgressBar pbLoadingDistricts;

    @BindView(R.id.rv_districts)
    RecyclerView rvDistricts;


    String sportSelected;
    List<ListItem> elementsList;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list);
        sportSelected = getIntent().getStringExtra(Constants.EXTRA_SPORT_SELECTED_NAME);
        String sportSelectedCount = getIntent().getStringExtra(Constants.EXTRA_COUNT);
        String subTitle = sportSelected + " (" + sportSelectedCount + ")";
        if (getSupportActionBar()!=null) {
            getSupportActionBar().setTitle(getString(R.string.title_district));
            getSupportActionBar().setSubtitle(subTitle);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        ButterKnife.bind(this);
        showLoading();
        List<Competition> competitions = CompetitionsDAO.queryCompetitionsBySport(this, sportSelected);
        elementsList = initElementsList(competitions);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        GenericAdapter genericAdapter = new GenericAdapter(this, this, elementsList);
        rvDistricts.setHasFixedSize(true);
        rvDistricts.setLayoutManager(layoutManager);
        rvDistricts.setAdapter(genericAdapter);
        genericAdapter.notifyDataSetChanged();
        hideLoading();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
            case R.id.action_home:
                Intent i=new Intent(this, MainActivity.class);
                i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(i);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private List<ListItem> initElementsList(List<Competition> competitions) {
        List<ListItem> listElements = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();
        for (Competition competition : competitions) {
            Integer count = map.get(competition.distrito());
            if (count==null) {
                count = 0;
            }
            map.put(competition.distrito(), count + 1);
        }
        for (String s : map.keySet()) {
            //String value = map.get(s).toString();
            ListItem listItem = new ListItem(s, map.get(s).toString());
            listElements.add(listItem);
        }
        Collections.sort(listElements, new ListItem.ListItemCompartor());
        return listElements;
    }

    private void hideLoading() {
        pbLoadingDistricts.setVisibility(View.INVISIBLE);
        rvDistricts.setVisibility(View.VISIBLE);
    }

    private void showLoading() {
        pbLoadingDistricts.setVisibility(View.VISIBLE);
        rvDistricts.setVisibility(View.INVISIBLE);
    }

    @Override
    public void onListItemClick(int clickedItemIndex) {
        Intent intent = new Intent(this, CategoriesActivity.class);
        String districtName = elementsList.get(clickedItemIndex).getName();
        String count = elementsList.get(clickedItemIndex).getCount();
        intent.putExtra(Constants.EXTRA_SPORT_SELECTED_NAME, sportSelected);
        intent.putExtra(Constants.EXTRA_DISTRICT_SELECTED_NAME, districtName);
        intent.putExtra(Constants.EXTRA_COUNT, count);
        startActivity(intent);
    }
}

File: app/src/main/java/com/adiaz/ligasmadrid/activities/DistrictActivity.kt
package com.adiaz.ligasmadrid.activities

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.adiaz.ligasmadrid.R
import com.adiaz.ligasmadrid.adapters.GenericAdapter
import com.adiaz.ligasmadrid.db.daos.GroupsDAO
import com.adiaz.ligasmadrid.db.entities.Group
import com.adiaz.ligasmadrid.utils.Constants
import com.adiaz.ligasmadrid.utils.ListItem
import com.adiaz.ligasmadrid.utils.UtilsPreferences
import kotlinx.android.synthetic.main.activity_list.*
import java.util.*

class DistrictActivity : AppCompatActivity(), GenericAdapter.ListItemClickListener {

    private var sportSelected: String? = null
    private var elementsList: List<ListItem>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list)
        sportSelected = intent.getStringExtra(Constants.EXTRA_SPORT_SELECTED_NAME)
        val sportSelectedCount = intent.getStringExtra(Constants.EXTRA_COUNT)
        var subTitle = sportSelected
        if (UtilsPreferences.isShowCompetitionsNumber(this)) {
            subTitle += " ($sportSelectedCount)"
        }
        setSupportActionBar(toolbar)
        if (supportActionBar != null) {
            supportActionBar!!.title = getString(R.string.title_district)
            supportActionBar!!.subtitle = subTitle
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }
        val groups = GroupsDAO.queryCompetitionsBySport(this, sportSelected!!)
        elementsList = initElementsList(groups)
        val layoutManager = LinearLayoutManager(this)
        val genericAdapter = GenericAdapter(this, this, elementsList)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = genericAdapter
        genericAdapter.notifyDataSetChanged()
        recyclerView.visibility = View.VISIBLE
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_home, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
            R.id.action_home -> {
                val i = Intent(this, MainActivity::class.java)
                i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                startActivity(i)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initElementsList(groups: List<Group>): List<ListItem> {
        val listElements = ArrayList<ListItem>()
        val map = HashMap<String, Int>()
        for (group in groups) {
            var count: Int? = map[group.distrito]
            if (count == null) {
                count = 0
            }
            map[group.distrito!!] = count + 1
        }
        for (s in map.keys) {
            //String value = map.get(s).toString();
            val listItem = ListItem(s, map[s].toString())
            listElements.add(listItem)
        }
        Collections.sort(listElements, ListItem.ListItemCompartor())
        return listElements
    }

    override fun onListItemClick(clickedItemIndex: Int) {
        val intent = Intent(this, CategoriesActivity::class.java)
        val districtName = elementsList!![clickedItemIndex].name
        val count = elementsList!![clickedItemIndex].count
        intent.putExtra(Constants.EXTRA_SPORT_SELECTED_NAME, sportSelected)
        intent.putExtra(Constants.EXTRA_DISTRICT_SELECTED_NAME, districtName)
        intent.putExtra(Constants.EXTRA_COUNT, count)
        startActivity(intent)
    }
}
